src/parser.js

```diff
@@ -29,75 +47,61 @@ const parseHeader = (header) => {
   const disposition = parseDisposition(header['content-disposition'][0])
 
   const details = type
-  details.name = disposition.name
+  details.name = decodeURIComponent(disposition.name)
   details.type = disposition.type
 
   return details
 }
 
-/**
- * Parser
- *
- * @constructor
- * @param {Object} options
- * @returns {Parser}
- */
-function Parser (options) {
-  // allow use without new
-  if (!(this instanceof Parser)) {
-    return new Parser(options)
-  }
-
-  this.dicer = new Dicer({ boundary: options.boundary })
-
-  this.dicer.on('part', (part) => this.handlePart(part))
-
-  this.dicer.on('error', (err) => this.emit('err', err))
-
-  this.dicer.on('finish', () => {
-    this.emit('finish')
-    this.emit('end')
-  })
-
-  Transform.call(this, options)
-}
-util.inherits(Parser, Transform)
-
-Parser.prototype._transform = function (chunk, enc, cb) {
-  this.dicer.write(chunk, enc)
-  cb()
-}
-
-Parser.prototype._flush = function (cb) {
-  this.dicer.end()
-  cb()
-}
-
-Parser.prototype.handlePart = function (part) {
-  part.on('header', (header) => {
-    const partHeader = parseHeader(header)
-
-    if (isDirectory(partHeader.mime)) {
-      part.on('data', () => false)
-      this.emit('directory', partHeader.name)
-      return
+async function * parser (boundary, stream) {
+  const dicer = new Dicer({ boundary })
+  stream.pipe(dicer)
+
+  for await (const part of subscribe(dicer, 'part', 'finish')) {
+    for await (const header of subscribe(part, 'header')) {
+      const partHeader = parseHeader(header)
+
+      if (isDirectory(partHeader.mime)) {
+        // consume data from stream so we move on to the next header/part
+        part.on('data', () => {})
+
+        yield {
+          type: 'directory',
+          name: partHeader.name
+        }
+
+        continue
+      }
+
+      if (partHeader.mime === applicationSymlink) {
+        // consume data from stream so we move on to the next header/part
+        part.on('data', () => {})
```

Comment:
Same here, should `part.on('data')` only be called once rather than for every header?

Reply:
I could be doing it wrong but in my testing I noticed that setting up a noop for the `data` event for all headers caused data to start being read before the calling code had a chance to set a listener up for data from `part.content`.

Reply:
Actually I think there might be a better way, will update this PR today or tomorrow.